pkg/ensurance/executor/evict.go

```diff
@@ -57,41 +48,76 @@ func (e *EvictExecutor) Avoid(ctx *ExecuteContext) error {
 	metrics.ExecutorStatusCounterInc(metrics.SubComponentEvict, metrics.StepAvoid)
 
-	var bSucceed = true
-	var errPodKeys []string
+	var errPodKeys, errKeys []string
+	// TODO: totalReleasedResource used for prom metrics
+	totalReleased := ReleaseResource{}
 
-	wg := sync.WaitGroup{}
+	/* The step to evict:
+	1. If EvictWaterLine has metrics that can't be quantified, select a evictable metric which has the highest action priority, use its EvictFunc to evict all selected pods, then return
+	2. Get the gaps between current usage and waterlines
+		2.1 If there is a metric that can't get current usage, select a evictable metric which has the highest action priority, use its EvictFunc to evict all selected pods, then return
+		2.2 Traverse metrics that can be quantified, if there is gap for the metric, then sort candidate pods by its SortFunc if exists, otherwise use GeneralSorter by default.
+	       Then evict sorted pods one by one util there is no gap to waterline
+	*/
 
-	for i := range e.EvictPods {
-		wg.Add(1)
+	metricsEvictQuantified, MetricsNotEvcitQuantified := e.EvictWaterLine.DivideMetricsByEvictQuantified()
 
-		go func(evictPod EvictPod) {
-			defer wg.Done()
+	// There is a metric that can't be EvictQuantified, so evict all selected pods
+	if len(MetricsNotEvcitQuantified) != 0 {
+		klog.V(6).Info("There is a metric that can't be EvcitQuantified")
 
-			pod, err := ctx.PodLister.Pods(evictPod.PodKey.Namespace).Get(evictPod.PodKey.Name)
-			if err != nil {
-				bSucceed = false
-				errPodKeys = append(errPodKeys, "not found ", evictPod.PodKey.String())
-				return
+		highestPrioriyMetric := e.EvictWaterLine.GetHighestPriorityEvictAbleMetric()
+		if highestPrioriyMetric != "" {
+			klog.V(6).Infof("The highestPrioriyMetric is %s", highestPrioriyMetric)
+			errPodKeys = e.evictPods(ctx, &totalReleased, highestPrioriyMetric)
+		}
+	} else {
+		_, _, ctx.EvictGapToWaterLines = buildGapToWaterLine(ctx.stateMap, ThrottleExecutor{}, *e, ctx.executeExcessPercent)
+
+		if ctx.EvictGapToWaterLines.HasUsageMissedMetric() {
+			klog.V(6).Infof("There is a metric usage missed")
+			highestPrioriyMetric := e.EvictWaterLine.GetHighestPriorityEvictAbleMetric()
+			if highestPrioriyMetric != "" {
+				errPodKeys = e.evictPods(ctx, &totalReleased, highestPrioriyMetric)
 			}
-
-			err = utils.EvictPodWithGracePeriod(ctx.Client, pod, evictPod.DeletionGracePeriodSeconds)
-			if err != nil {
-				bSucceed = false
-				errPodKeys = append(errPodKeys, "evict failed ", evictPod.PodKey.String())
-				klog.Warningf("Failed to evict pod %s: %v", evictPod.PodKey.String(), err)
-				return
+		} else {
+			// The metrics in EvictGapToWaterLines are can be EvictQuantified and has current usage, then evict precisely
+			var released ReleaseResource
+			wg := sync.WaitGroup{}
+			for _, m := range metricsEvictQuantified {
+				klog.V(6).Infof("Evict precisely on metric %s", m)
+				if MetricMap[m].SortAble {
```

Comment:
if the MetricMap is only used in this package, it should use lowercase in begining